ocaml/xapi/xapi_clustering.ml

```diff
@@ -401,3 +424,99 @@ let compute_corosync_max_host_failures ~__context =
+     new state"
+    __FUNCTION__ (List.length updates) ;
+  let m =
+    Cluster_client.LocalClient.diagnostics (rpc ~__context)
+      "update quorum api fields with diagnostics"
+  in
+  match Idl.IdM.run @@ Cluster_client.IDL.T.get m with
+  | Ok diag ->
+      Db.Cluster.set_is_quorate ~__context ~self:cluster ~value:diag.is_quorate ;
+      let all_cluster_hosts = Db.Cluster_host.get_all ~__context in
+      let current_time = API.Date.now () in
+      ( match diag.quorum_members with
+      | None ->
+          List.iter
+            (fun self ->
+              Db.Cluster_host.set_live ~__context ~self ~value:false ;
+              Db.Cluster_host.set_last_update_live ~__context ~self
+                ~value:current_time
+            )
+            all_cluster_hosts
+      | Some nodel ->
+          let quorum_hosts =
+            List.map
+              (fun {addr= IPv4 addr; _} ->
+                match find_cluster_host_by_network ~__context ~addr with
+                | None ->
+                    error "cannot find cluster host with network address %s"
+                      addr ;
+                    handle_error
+                      (InternalError "cannot find host for network address")
+                | Some ch ->
+                    ch
+              )
+              nodel
+          in
+
+          let missing_hosts =
+            List.filter
+              (fun h -> not (List.mem h quorum_hosts))
+              all_cluster_hosts
+          in
+          List.iter
+            (fun self ->
+              Db.Cluster_host.set_live ~__context ~self ~value:true ;
+              Db.Cluster_host.set_last_update_live ~__context ~self
+                ~value:current_time
+            )
+            quorum_hosts ;
+          List.iter
+            (fun self ->
+              Db.Cluster_host.set_live ~__context ~self ~value:false ;
+              Db.Cluster_host.set_last_update_live ~__context ~self
+                ~value:current_time
+            )
+            missing_hosts
+      ) ;
+      Db.Cluster.set_quorum ~__context ~self:cluster
+        ~value:(Int64.of_int diag.quorum) ;
+      Db.Cluster.set_total_votes ~__context ~self:cluster
+        ~value:(Int64.of_int diag.total_votes)
+  | Error error ->
+      warn "cannot query diagnostics" ;
+      handle_error error
+
+let create_cluster_watcher_on_master ~__context ~host =
+  if Helpers.is_pool_master ~__context ~host then (
+    let watch () =
+      while true do
+        let m =
+          Cluster_client.LocalClient.UPDATES.get (rpc ~__context)
+            "call cluster watcher" 3.
+        in
+        match Idl.IdM.run @@ Cluster_client.IDL.T.get m with
+        | Ok updates -> (
+          match find_cluster_host ~__context ~host with
+          | Some ch ->
+              let cluster = Db.Cluster_host.get_cluster ~__context ~self:ch in
+              on_corosync_update ~__context ~cluster updates
+          | None ->
+              ()
```

Comment:
Does this mean the current coordinator is not in cluster?

Reply:
possibly, in which case doing nothing could be a sensible choice, I suppose...